### ISSUE TITLE
RFC: Official Race and its requirements

### ISSUE THREAD

**Issue by axblk** [teeworlds/teeworlds#2789]
I recently talked with @Zatline about finally adding a Race gamemode.
Basically there are two variants of doing this: a very simple round based gametype like proposed here #33 or something ladder-based, where you just aim for the fastest time. While the first approach fits better with the simplistic design of the game, Race in Teeworlds has been associated with a ladder-based system for over a decade now.
And just to be realistic: a round-based game mode is only fun if there are enough players. Considering today's player-count there will hardly ever be a real game.

So what do you think about adding a simple ladder-based solo racing mode as official Race in Teeworlds? To be clear: I don't mean anything as complex as ddrace/ddnet.

#### Per-map ranking
For each map the time for each player is saved. There should be a method to show the ladder for the current map and to query the time and rank for individual players. Basically any racing mod out there uses `/top5` and `/rank` for this. These commands could be added as well, but the main method to query and display these information should be a simple UI in the ingame menu.

Probably the biggest issue here is how to save the these times. The best approach would be to have a local database (for example using sqlite), a server/client based database (like MySQL) is definitely out of scope here.
The question is still: do we want an additional (pretty huge) library just to store times in a single gametype?
The alternative would be to store the times in a simple text or json file. The main problem here is that we have to recreate the entire file after each finish. It works for a small number of players but at some point that can definitely be a limiting factor.

#### Tiles and Layers
Racing obviously requires tiles for the start and end, adding them is kinda easy.
For creating interesting race maps one definitely wants to have some other new tiles. Simple ones: one-way, hook-through. More complex ones: teleporters, speedup, checkpoints. More suggestions are welcome, but keep it simple! In my opinion, interactive tiles that have state would be too much here. Of course, it won't work for old clients, but all new tiles should be predictable.

Adding new simple tiles is not a big deal, but things like teleporters are not possible with the current game layer. They require additional parameters per tile, for example teleporters need a way to connect them to each other.
Some basic thoughts about that: allow having multiple game layers (for example water requires this as well), for each of these layers one can define a certain number of additional per-tile parameters (1, 2, 4 bytes each?). Teleports would for example require one parameter to connect them. Mods could reuse this so they don't need to hard-code additional layer types anymore. This idea is still missing a lot of details, like for example the editor integration. More on that in a new issues.

#### Map-Settings
For racing per-map settings would be useful. For example for turning off damage or enabling infinite ammo. The maps should be self-contained, so the settings should be embedded into the map file and in my opinion it should only be possible to modify them in the editor and not via rcon.

#### Effects, Graphics, Details
One thing all the racing game modes out there are missing is proper details. They all feel raw and unpolished. Just to name a few examples: sounds and effects for start, finish, teleporters and speedups.

#### Standard maps
Like all other official game modes there should be a few official maps. These maps should have a nice tee-ish design.

#### Ideas, recommendations, ...?
The major blockers are probably storage of times (plain file vs database) and extended game layers.
What do you think about all this? Any other ideas or things that could be improved? Did I completely miss something?

Note about ddnet/ddrace: I'd like to have a design that makes it easy to base something like ddrace on it, BUT I do not aim for any compatibility with the current implementations. Their scoring system is way beyond anything we would ever need in Teeworlds, the physics code is a lot more complex and buggy (due to historic reasons) and the layer implementation is neither flexible nor extendable. One could still create a tool to convert maps.

**Comment by l-ouis** [comment]
inb4 dev response:
- doesnt fit teeworlds

but i think this would be a very good idea. it would introduce a mode different enough from ctf and dm that it won't confuse players on how physics/mechanics work and would also be very fun and beneficial for the game. it would also help newbies learn race and possibly have a better experience on more popular servers like ddnet. 

however, there are some issues - first off, sounds for start, teleporters, and speedups would be annoying tbh. a sound for just finish would be better (or maybe a small non-intrusive sound for the startline). Also, i'm not sure about the idea of speedups. I feel like a big aspect of teeworlds is that you are in control of your own movement, and speedups don't really feel satisfying as just landing a good swing to build momentum.

also, the issue of maps is meh. the enjoying thing about modes like ddrace is the variety of different maps that you can play. if there are official race maps, they must be well made, meaning that it won't be possible to have a wide variety of race maps to play. this would make the game boring for some people after they finish most of the race maps.

**Comment by ChillerDragon** [comment]
Im sure modders will be happy if sqlite3 is already in the code base.

**Comment by heinrich5991** [comment]
> The question is still: do we want an additional (pretty huge) library just to store times in a single gametype?

Yes, we should probably use sqlite3 for this purpose. Its size is apparently around 700KB (according to https://en.wikipedia.org/wiki/SQLite) and it handles all the synchronization and partial file write issues for us, so we won't corrupt the race scores on unclean exits or when multiple servers are using the same DB.

**Comment by Dune-jr** [comment]
Related: https://github.com/teeworlds/teeworlds/issues/2172

Sounds good to me. I really like the idea of official race maps, and maps that cannot be tuned other than in the editor - this means quite some client-sided work though. I'm also not sure if we should support infinite ammo.

There were a lot of recurring discussions back in 0.6 to add a race mod, but iirc the general consensus back then was that no matter what official Teeworlds does, it would not compete with teerace, so there was little point to do that.
Lack of any compatibility with the way the existing race maps are implemented would probably make the game mode somewhat unpopular because you would need to convert maps (people struggled a lot with forks of @axblk's race in 0.7 because of that).

**Comment by jxsl13** [comment]
@ChillerDragon Agreed.
SQLite is perfect for that purpose.

**Comment by oy** [comment]
@Dune-jr wrote:
> 
> There were a lot of recurring discussions back in 0.6 to add a race mod, but iirc the general consensus back then was that no matter what official Teeworlds does, it would not compete with teerace, so there was little point to do that.

Yeah, that's why we didn't add an official race mode. And it differs quite much from the other official game types and has different requirements.

About race types:
There used to be another type of race where all players started at the same time and the one who reached the end first won.

**Comment by Dune-jr** [comment]
> There used to be another type of race where all players started at the same time and the one who reached the end first won.

Implementing this sounds interesting and maybe very close to the spirit of what currently is Teeworlds

**Comment by Fisico** [comment]
> > There used to be another type of race where all players started at the same time and the one who reached the end first won.
> 
> Implementing this sounds interesting and maybe very close to the spirit of what currently is Teeworlds

It sounds good on paper but as @axblk already mentioned, with the little player base we will never see full servers because the people to race against will be missing. Also for new players it is impossible to win a race map against an experienced player. With a solo race mode it is mostly against yourself and shaving off time over the many hundred tries you make, like in trackmania.

A single player based racemod fits teeworlds very well and got well established over the last 10 years. Infinite ammo is not a problem at all because this is a completely different gametype compared to ctf/dm, everyone will figure that out in a few seconds. Maybe it is also a trigger to add an ammo display under the health UI. Long overdue for survival for example.

A nice UI that displays the sector times and the top10 and some polishing is all, a new racemod needs from a player perspective.

It needs some more sounds, but I’m not sure if we/I can design a subtle enough teleport sound, that does not get annoying fast. I did most of them already so I will also try to make the others happen.

I would add a few new things to time taking:
Personal Best: green color, compare each sector/checkpoint.
Server Best: purple color, compare each sector/checkpoint to server best, if the sector/checkpoint is faster than the server/sector best.

I think it is time to give something completely new a shot. It has a lot of potential, if the player base does not accept it, it will not kill teeworlds, it “only” took time to implement it. Maybe it will help mods and this will keep teeworlds alive who knows. I can’t see anything wrong with at least trying it.
